quil-rs/src/expression/simplification/by_hand.rs

```diff
@@ -0,0 +1,830 @@
+            Expression::Number(x * y)
+        }
+        (Expression::Number(x), InfixOperator::Star, Expression::PiConstant) => {
+            Expression::Number(x * PI)
+        }
+        (Expression::PiConstant, InfixOperator::Star, Expression::Number(y)) => {
+            Expression::Number(PI * y)
+        }
+        (Expression::PiConstant, InfixOperator::Star, Expression::PiConstant) => {
+            Expression::Number(PI * PI)
+        }
+
+        // Division with zero
+        (Expression::Number(x), InfixOperator::Slash, _) if is_zero(x) => Expression::Number(ZERO),
+        (_, InfixOperator::Slash, Expression::Number(y)) if is_zero(y) => {
+            Expression::Number(real!(f64::NAN)) // TODO Is this OK?
+        }
+        // Division with one
+        (left, InfixOperator::Slash, Expression::Number(y)) if is_one(y) => left,
+        // Division with self
+        (left, InfixOperator::Slash, right) if left == right => Expression::Number(ONE),
+        // Division with numbers or π (π / π already covered)
+        (Expression::Number(x), InfixOperator::Slash, Expression::Number(y)) => {
+            Expression::Number(x / y)
+        }
+        (Expression::Number(x), InfixOperator::Slash, Expression::PiConstant) => {
+            Expression::Number(x / PI)
+        }
+        (Expression::PiConstant, InfixOperator::Slash, Expression::Number(y)) => {
+            Expression::Number(PI / y)
+        }
+
+        // Exponentiation
+
+        // Exponentiation with zero
+        (Expression::Number(x), InfixOperator::Caret, _) if is_zero(x) => Expression::Number(ZERO),
+        (_, InfixOperator::Caret, Expression::Number(y)) if is_zero(y) => Expression::Number(ONE),
+        // Exponentiation with one
+        (Expression::Number(x), InfixOperator::Caret, _) if is_one(x) => Expression::Number(ONE),
+        (left, InfixOperator::Caret, Expression::Number(y)) if is_one(y) => left,
+
+        //----------------------------------------------------------------
+        // Next: dealing with negation in subexpressions
+        //----------------------------------------------------------------
+
+        // Addition with negation
+        (
+            ref left,
+            InfixOperator::Plus,
+            Expression::Prefix(PrefixExpression {
+                operator: PrefixOperator::Minus,
+                ref expression,
+            }),
+        ) => simplify_infix(
+            left,
+            InfixOperator::Minus,
+            expression,
+            limit.saturating_sub(1),
+        ),
+        (
+            Expression::Prefix(PrefixExpression {
+                operator: PrefixOperator::Minus,
+                ref expression,
+            }),
+            InfixOperator::Plus,
+            ref right,
+        ) => simplify_infix(
+            right,
+            InfixOperator::Minus,
+            expression,
+            limit.saturating_sub(1),
+        ),
```

Comment:
Not sure about the syntax with this one, but you get the idea

```suggestion
        (
            ref left: other,
            InfixOperator::Plus,
            Expression::Prefix(PrefixExpression {
                operator: PrefixOperator::Minus,
                ref expression: inner,
            }),
        ) | (
            Expression::Prefix(PrefixExpression {
                operator: PrefixOperator::Minus,
                ref expression: inner,
            }),
            InfixOperator::Plus,
            ref right: other,
        ) => simplify_infix(
            other,
            InfixOperator::Minus,
            inner,
            limit.saturating_sub(1),
        ),
```